pkg/etcd/etcd.go

```diff
@@ -1067,15 +1077,33 @@ func (e *ETCD) manageLearners(ctx context.Context) {
 			logrus.Warnf("Failed to list nodes with etcd role: %v", err)
 		}
 
+		// a map to track if a node is a member of the etcd cluster or not
+		nodeIsMember := make(map[string]bool)
+		nodesMap := make(map[string]*v1.Node)
+		for _, node := range nodes {
+			nodeIsMember[node.Name] = false
+			nodesMap[node.Name] = node
+		}
+
 		for _, member := range members.Members {
 			if member.IsLearner {
 				if err := e.trackLearnerProgress(ctx, progress, member); err != nil {
 					logrus.Errorf("Failed to track learner progress towards promotion: %v", err)
 				}
 				for _, node := range nodes {
 					if strings.HasPrefix(member.Name, node.Name+"-") {
-						_, _, err := e.setEtcdStatusCondition(node, client.CoreV1(), member.Name, false)
-						if err != nil {
+						nodeIsMember[node.Name] = true
+
+						// verify if the member is healthy and set the etcd status condition
+						if _, err := e.client.Status(ctx, member.ClientURLs[0]); err != nil {
```

Comment:
You might need to do this with a specific per-member context with a shorter timeout. Otherwise a single inaccessible member could hang the entire loop.

You might also see if any of the code here overlaps with the status retrieved in `trackLearnerProgress`; if I remember correctly that also makes a Status call that you could perhaps return for this status check.

Reply:
I was thinking of putting it in the `trackLearnerProgress`, but the scope might get kind of big. Still, we can handle it and set the status in the `trackLearnerProgress` error.